File: src/pal/ollama_provider.py
"""
Ollama provider implementation for local LLM inference.

Reference: docs/keikaku.md Section 4.1 - Provider Abstraction Layer
"""

import json
import time
from typing import Any, Dict, Iterator, List, Optional

import httpx

from .base import BaseProvider, ProviderCapabilities, ProviderFactory


class OllamaProvider(BaseProvider):
    """
    Provider for Ollama local LLM server.
    
    Supports any model running on Ollama (Qwen3, Llama, etc.).
    """
    
    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        self.base_url = config.get("base_url", "http://localhost:11434")
        self.api_generate = f"{self.base_url}/api/generate"
        self.api_chat = f"{self.base_url}/api/chat"
        self.api_tags = f"{self.base_url}/api/tags"
        
        # HTTP client with timeout
        self.client = httpx.Client(timeout=self.timeout)
    
    def generate(
        self,
        messages: List[Dict[str, str]],
        params: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Generate text using Ollama chat API.
        
        Args:
            messages: List of {'role': 'user'|'assistant'|'system', 'content': str}.
            params: Generation parameters.
        
        Returns:
            Generated text.
        """
        params = params or {}
        
        # Convert messages to Ollama format
        ollama_messages = self._convert_messages(messages)
        
        # Build request payload
        payload = {
            "model": self.model,
            "messages": ollama_messages,
            "stream": False,
            "options": {
                "temperature": params.get("temperature", 0.7),
                "num_predict": params.get("max_tokens", 2000),
                "top_p": params.get("top_p", 0.9),
            }
        }
        
        # Add system message if present
        system_content = self._extract_system_message(messages)
        if system_content:
            payload["system"] = system_content
        
        try:
            response = self.client.post(self.api_chat, json=payload)
            response.raise_for_status()
            
            data = response.json()
            return data.get("message", {}).get("content", "")
        
        except httpx.HTTPError as e:
            raise RuntimeError(f"Ollama API error: {e}")
    
    def generate_stream(
        self,
        messages: List[Dict[str, str]],
        params: Optional[Dict[str, Any]] = None
    ) -> Iterator[str]:
        """
        Generate text with streaming.
        
        Yields text chunks as they are generated.
        """
        params = params or {}
        
        ollama_messages = self._convert_messages(messages)
        
        payload = {
            "model": self.model,
            "messages": ollama_messages,
            "stream": True,
            "options": {
                "temperature": params.get("temperature", 0.7),
                "num_predict": params.get("max_tokens", 2000),
                "top_p": params.get("top_p", 0.9),
            }
        }
        
        try:
            with self.client.stream("POST", self.api_chat, json=payload) as response:
                response.raise_for_status()
                
                for line in response.iter_lines():
                    if line:
                        try:
                            data = json.loads(line)
                            chunk = data.get("message", {}).get("content", "")
                            if chunk:
                                yield chunk
                            
                            # Check if done
                            if data.get("done", False):
                                break
                        except json.JSONDecodeError:
                            continue
        
        except httpx.HTTPError as e:
            raise RuntimeError(f"Ollama streaming error: {e}")
    
    def capabilities(self) -> ProviderCapabilities:
        """
        Report Ollama capabilities.
        
        Note: Actual capabilities depend on the loaded model.
        """
        # Default capabilities for Qwen3-1.7B
        return ProviderCapabilities(
            ctx_len=32768,  # Qwen3-1.7B context length
            supports_tools=False,  # Ollama may support tools depending on model
            supports_json_mode=False,  # JSON mode depends on model
            supports_thinking_mode=True,
            supports_streaming=True,
        )
    
    def healthcheck(self) -> bool:
        """
        Check if Ollama server is running.
        
        Returns:
            True if server is reachable.
        """
        try:
            response = self.client.get(self.api_tags, timeout=5)
            if response.status_code == 200:
                # Check if our model is available
                data = response.json()
                models = [m.get("name", "") for m in data.get("models", [])]
                return self.model in models or any(self.model in m for m in models)
            return False
        except httpx.RequestError:
            return False
    
    def _convert_messages(self, messages: List[Dict[str, str]]) -> List[Dict[str, str]]:
        """
        Convert standard messages to Ollama format.
        
        Ollama chat API expects: [{"role": "user"|"assistant", "content": "..."}]
        """
        converted = []
        for msg in messages:
            role = msg.get("role", "user")
            content = msg.get("content", "")
            
            # Skip system messages (handled separately in Ollama)
            if role == "system":
                continue
            
            converted.append({
                "role": role,
                "content": content
            })
        
        return converted
    
    def _extract_system_message(self, messages: List[Dict[str, str]]) -> Optional[str]:
        """Extract system message content if present."""
        for msg in messages:
            if msg.get("role") == "system":
                return msg.get("content")
        return None
    
    def list_models(self) -> List[str]:
        """List available models on Ollama server."""
        try:
            response = self.client.get(self.api_tags)
            response.raise_for_status()
            data = response.json()
            return [m.get("name", "") for m in data.get("models", [])]
        except httpx.HTTPError:
            return []


# Register provider
ProviderFactory.register("ollama", OllamaProvider)

File: go/pkg/agents/ollama_provider.go
package agents

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/novelist/novelist/pkg/models"
)

const defaultOllamaBaseURL = "http://localhost:11434"

type ollamaProvider struct {
	model   string
	baseURL string
	client  *http.Client
}

type ollamaChatRequest struct {
	Model    string                 `json:"model"`
	Messages []Message              `json:"messages"`
	Stream   bool                   `json:"stream"`
	Format   string                 `json:"format,omitempty"`
	Options  map[string]interface{} `json:"options,omitempty"`
}

type ollamaChatResponse struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
	PromptEvalCount int    `json:"prompt_eval_count"`
	EvalCount       int    `json:"eval_count"`
	Error           string `json:"error"`
}

func init() {
	RegisterProviderFactory("ollama", NewOllamaProvider)
}

// NewOllamaProvider creates a provider backed by Ollama HTTP API.
func NewOllamaProvider(config models.ProviderConfig) (Provider, error) {
	if strings.TrimSpace(config.Model) == "" {
		return nil, fmt.Errorf("ollama provider requires model")
	}

	baseURL := strings.TrimRight(strings.TrimSpace(config.BaseURL), "/")
	if baseURL == "" {
		baseURL = defaultOllamaBaseURL
	}

	timeout := time.Duration(config.Timeout) * time.Second
	if timeout <= 0 {
		timeout = 120 * time.Second
	}

	return &ollamaProvider{
		model:   config.Model,
		baseURL: baseURL,
		client:  &http.Client{Timeout: timeout},
	}, nil
}

func (p *ollamaProvider) Name() string {
	return "ollama"
}

func (p *ollamaProvider) Generate(ctx context.Context, messages []Message, params GenerateParams) (*models.GenerationResult, error) {
	reqPayload := ollamaChatRequest{
		Model:    p.model,
		Messages: messages,
		Stream:   false,
		Options: map[string]interface{}{
			"temperature": params.Temperature,
			"num_predict": params.MaxTokens,
			"top_p":       params.TopP,
		},
	}
	if params.JSONMode {
		reqPayload.Format = "json"
	}

	body, err := json.Marshal(reqPayload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal ollama request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.baseURL+"/api/chat", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to build ollama request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ollama request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= http.StatusBadRequest {
		raw, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return nil, fmt.Errorf("ollama returned %d: %s", resp.StatusCode, strings.TrimSpace(string(raw)))
	}

	var out ollamaChatResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, fmt.Errorf("failed to decode ollama response: %w", err)
	}
	if out.Error != "" {
		return nil, fmt.Errorf("ollama error: %s", out.Error)
	}

	promptTokens := out.PromptEvalCount
	completionTokens := out.EvalCount
	if promptTokens <= 0 {
		promptTokens = estimateTokensFromMessages(messages)
	}
	if completionTokens <= 0 {
		completionTokens = estimateTokensFromText(out.Message.Content)
	}

	return &models.GenerationResult{
		Text:             strings.TrimSpace(out.Message.Content),
		PromptTokens:     promptTokens,
		CompletionTokens: completionTokens,
	}, nil
}

func (p *ollamaProvider) Capabilities() ProviderCapabilities {
	return ProviderCapabilities{
		CtxLen:            32768,
		SupportsTools:     false,
		SupportsJSONMode:  true,
		SupportsThinking:  true,
		SupportsStreaming: false,
	}
}

func (p *ollamaProvider) HealthCheck(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.baseURL+"/api/tags", nil)
	if err != nil {
		return err
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("status %d", resp.StatusCode)
	}
	return nil
}

func estimateTokensFromMessages(messages []Message) int {
	total := 0
	for _, msg := range messages {
		total += estimateTokensFromText(msg.Content)
	}
	if total <= 0 {
		return 1
	}
	return total
}

func estimateTokensFromText(text string) int {
	if text == "" {
		return 1
	}
	chars := len([]rune(text))
	estimated := chars / 4
	if estimated <= 0 {
		return 1
	}
	return estimated
}
